patches/tModLoader/Terraria/ModLoader/Core/AssemblyManager.cs

```diff
@@ -339,18 +339,29 @@ internal static bool FirstModInStackTrace(StackTrace stack, out string modName)
 
 	private static bool IsLoadable(ModLoadContext mod, Type type)
 	{
-		foreach (var attr in type.GetCustomAttributesData()) {
-			if (attr.AttributeType.AssemblyQualifiedName == typeof(ExtendsFromModAttribute).AssemblyQualifiedName) {
-				var modNames = (IEnumerable<CustomAttributeTypedArgument>)attr.ConstructorArguments[0].Value;
-				if (!modNames.All(v => mod.IsModDependencyPresent((string)v.Value)))
-					return false;
+		try {
+			foreach (var attr in type.GetCustomAttributesData()) {
+				if (attr.AttributeType.AssemblyQualifiedName == typeof(ExtendsFromModAttribute).AssemblyQualifiedName) {
+					var modNames = (IEnumerable<CustomAttributeTypedArgument>)attr.ConstructorArguments[0].Value;
+					if (!modNames.All(v => mod.IsModDependencyPresent((string)v.Value)))
+						return false;
+				}
 			}
-		}
 
-		if (type.BaseType != null && !IsLoadable(mod, type.BaseType))
-			return false;
+			if (type.BaseType != null && !IsLoadable(mod, type.BaseType))
+				return false;
+
+			if (type.DeclaringType != null && !IsLoadable(mod, type.DeclaringType))
+				return false;
 
-		return type.GetInterfaces().All(i => IsLoadable(mod, i));
+			return type.GetInterfaces().All(i => IsLoadable(mod, i));
+		}
+		catch (FileNotFoundException e) {
+			throw new Exceptions.GetLoadableTypesException(
+				$"This mod seems to inherit from classes in another mod. Use the [ExtendsFromMod] attribute to allow this mod to load when that mod is not enabled.\n\nThe \"{type.FullName}\" class caused this error." + "\n\n" + e.Message,
+				e
+			);
```

Comment:
Less complicated code/error message duplication if you do. Then you can just conditionally add the `The \"{type.FullName}\" class caused this error.` line if `Data["type"] is Type t`
```cs
catch (Exception e) {
	e.Data["type"] = type;
	throw;
}
```